api/openai/files.go

```diff
@@ -0,0 +1,186 @@
+package openai
+
+import (
+	"fmt"
+	config "github.com/go-skynet/LocalAI/api/config"
+	"github.com/go-skynet/LocalAI/api/options"
+	"github.com/gofiber/fiber/v2"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+var uploadedFiles []File
+
+// File represents the structure of a file object from the OpenAI API.
+type File struct {
+	ID        string    `json:"id"`         // Unique identifier for the file
+	Object    string    `json:"object"`     // Type of the object (e.g., "file")
+	Bytes     int       `json:"bytes"`      // Size of the file in bytes
+	CreatedAt time.Time `json:"created_at"` // The time at which the file was created
+	Filename  string    `json:"filename"`   // The name of the file
+	Purpose   string    `json:"purpose"`    // The purpose of the file (e.g., "fine-tune", "classifications", etc.)
+}
+
+// UploadFilesEndpoint https://platform.openai.com/docs/api-reference/files/create
+func UploadFilesEndpoint(cm *config.ConfigLoader, o *options.Option) func(c *fiber.Ctx) error {
+	return func(c *fiber.Ctx) error {
+		file, err := c.FormFile("file")
+		if err != nil {
+			return err
+		}
+
+		// Check the file size
+		if file.Size > int64(o.UploadLimitMB) {
+			return c.Status(fiber.StatusBadRequest).SendString(fmt.Sprintf("File size %d exceeds upload limit %d", file.Size, o.UploadLimitMB))
+		}
+
+		purpose := c.FormValue("purpose", "") //TODO put in purpose dirs
+		if purpose == "" {
+			return c.Status(fiber.StatusBadRequest).SendString("Purpose is not defined")
+		}
+
+		// Sanitize the filename to prevent directory traversal
+		filename := sanitizeFileName(file.Filename)
+
+		// Create the directory if it doesn't exist
+		err = os.MkdirAll(o.UploadDir, os.ModePerm)
+		if err != nil {
+			return c.Status(fiber.StatusInternalServerError).SendString("Failed to create directory: " + err.Error())
+		}
+
+		savePath := filepath.Join(o.UploadDir, filename)
+
+		err = c.SaveFile(file, savePath)
+		if err != nil {
+			return c.Status(fiber.StatusInternalServerError).SendString("Failed to save file: " + err.Error())
+		}
+
+		f := File{
+			ID:        fmt.Sprintf("file-%d", time.Now().Unix()),
+			Object:    "file",
+			Bytes:     int(file.Size),
+			CreatedAt: time.Now(),
+			Filename:  file.Filename,
+			Purpose:   purpose,
+		}
+
+		uploadedFiles = append(uploadedFiles, f)
+
+		return c.Status(fiber.StatusOK).JSON(f)
+	}
+}
+
+func sanitizeFileName(fileName string) string {
```

Comment:
We do sanitize in https://github.com/mudler/LocalAI/blob/2e61ff32adf249955225ac3e50842c35ac17beea/pkg/utils/path.go#L19, but this implementation is slightly different (which I'm totally fine with)

Maybe we could move that under pkg/utils/path.go to be consistent.